internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,85 @@
+package logstate
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/plan"
+)
+
+var logBackends = map[string]func(*plan.LogTarget) (LogBackend, error){}
+
+func RegisterLogBackend(name string, builder func(*plan.LogTarget) (LogBackend, error)) {
+	validator := func(t *plan.LogTarget) error {
+		backend, err := builder(t)
+		if backend != nil {
+			backend.Close()
+		}
+		return err
+	}
+	plan.RegisterLogBackend(name, validator)
+}
+
+type LogManager struct {
+	mutex   sync.Mutex
+	targets map[string]*LogTarget
+	// targetsByService maps service name to the list of targets to receive its logs.
+	targetsByService map[string][]*LogTarget
+	// targetSelection maps log target names to selection criteria.
+	targetSelection map[string]string
+}
+
+func NewLogManager() *LogManager {
+	return &LogManager{
+		targets:          make(map[string]*LogTarget),
+		targetsByService: make(map[string][]*LogTarget),
+		targetSelection:  make(map[string]string),
+	}
+}
+
+func (m *LogManager) GetTargets(serviceName string) ([]*LogTarget, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	targets, ok := m.targetsByService[serviceName]
+	if !ok {
+		return nil, fmt.Errorf("no known logging targets for service %q", serviceName)
+	}
+	return targets, nil
+}
+
+// PlanChanged handles updates to the plan (server configuration),
+// stopping the previous checks and starting the new ones as required.
+func (m *LogManager) PlanChanged(p *plan.Plan) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if len(p.LogTargets) == 0 {
+		return
+	}
+
+	// update targets
+	for name, target := range p.LogTargets {
```

Comment:
I sort of alluded to this above but I don't think we should be trying to swap backends on a target.

My personal opinion is that we should just flush the current target of all log messages and stand up a new one with a new backend. From a concurrency point of view it's moves some of the locking back here to the manager on a re-plan that happens much more infrequently.

Reply:
Alternatively we could pinch all of the unsent messages on the current target and move them to the new target as well if flushing to the old target wasn't an option.